File: algorithm/002-add_two_numbers/kotlin/solution.kt
class ListNode(var `val` : Int = 0) {
    var next: ListNode? = null
}

class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var sum : ListNode = ListNode(0)
        var head = sum
        var carry = 0

        var p = l1
        var q = l2
        while (p != null || q != null || carry > 0) {
            if (p != null) {
                sum.`val` += p.`val`
                p = p.next
            }

            if (q != null) {
                sum.`val` += q.`val`
                q = q.next
            }

            sum.`val` += carry
            carry = sum.`val` / 10
            sum .`val` %= 10

            if (p != null || q != null || carry > 0) {
                sum.next = ListNode(0)
                sum = sum.next!!
            }
        }

        return head
    }
}

File: algorithm/002-add_two_numbers/go/solution.go
package solution

// ListNode singly-linked
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	sum := &ListNode{0, nil}
	head := sum
	carry := 0

	for l1 != nil || l2 != nil || carry > 0 {
		if l1 != nil {
			sum.Val += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			sum.Val += l2.Val
			l2 = l2.Next
		}

		sum.Val += carry
		carry = sum.Val / 10
		sum.Val %= 10

		if l1 != nil || l2 != nil || carry > 0 {
			sum.Next = &ListNode{0, nil}
			sum = sum.Next
		}
	}

	return head
}
